HW2/P2/P2.py

```diff
@@ -56,27 +78,49 @@
     dest = dest.astype(np.int32)
 
     # serial move
-    counts[:] = orig_counts
-    with Timer() as t:
-        move_data_serial(counts, src, dest, 100)
-    assert counts.sum() == total, "Wrong total after move_data_serial"
-    print("Serial correlated: {} seconds".format(t.interval))
+    times = []
+    for i in range(1):
+        counts[:] = orig_counts
+        with Timer() as t:
+            move_data_serial(counts, src, dest, 100)
+        assert counts.sum() == total, "Wrong total after move_data_serial"
+
+        current_time = t.interval
+        print("Serial correlated: {} seconds".format(current_time))
+        times.append(current_time)
+
+    print "Average", np.mean(times), "\n\n"
+
     serial_counts = counts.copy()
 
     # fine grained
-    counts[:] = orig_counts
-    with Timer() as t:
-        move_data_fine_grained(counts, src, dest, 100)
-    assert counts.sum() == total, "Wrong total after move_data_fine_grained"
-    print("Fine grained correlated: {} seconds".format(t.interval))
+    times = []
+    for i in range(1):
+        counts[:] = orig_counts
+        with Timer() as t:
+            move_data_fine_grained(counts, src, dest, 100)
+        assert counts.sum() == total, "Wrong total after move_data_fine_grained"
+
+        current_time = t.interval
+        print("Fine grained correlated: {} seconds".format(current_time))
+        times.append(current_time)
+
+    print "Average", np.mean(times), "\n\n"
 
     ########################################
     # You should explore different values for the number of locks in the medium
     # grained locking
     ########################################
     N = 10
-    counts[:] = orig_counts
-    with Timer() as t:
-        move_data_medium_grained(counts, src, dest, 100, N)
-    assert counts.sum() == total, "Wrong total after move_data_medium_grained"
-    print("Medium grained correlated: {} seconds".format(t.interval))
+    times = []
+    for N in [1,5,10,20,50,100]:
+        counts[:] = orig_counts
+        with Timer() as t:
+            move_data_medium_grained(counts, src, dest, 100, N)
+        assert counts.sum() == total, "Wrong total after move_data_medium_grained"
+
+        current_time = t.interval
+        print("Medium grained correlated with {} locks: {} seconds".format(N, current_time))
+        times.append(current_time)
+
```

Comment:
Same here, I know that it is clear that the serial does much better, but if you plot these results overlaid with the uncorrelated data you may see some interesting things, or at least I did. :)